Add S3ToRedshift example dag and system test

feluelle opened the pull request:
- add howto docs for S3ToRedshift example dag
- add terraform which runs terraform CLI commands in an isolated docker container

NOTE: This system test uses terraform to provide the infrastructure needed to run this example dag.

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

feluelle commented on tests/test_utils/postgres_system_helpers.py:
```diff
@@ -0,0 +1,97 @@
+        from os import getenv
+        connection_string = getenv("AIRFLOW_CONN_POSTGRES_DEFAULT")
+        auth = ["psql", connection_string]
+        executor.execute_cmd(cmd=auth + ["-c", cmd])
```
> from os import getenv
> connection_string = getenv("AIRFLOW_CONN_POSTGRES_DEFAULT")

Getting the connection through a env variable that needs to be defined :/ Is there a better way to do that?

> auth = ["psql", connection_string]
> executor.execute_cmd(cmd=auth + ["-c", cmd])

I don't like that part. I want to authenticate the context / the `psql` commands - that was the only way it worked. I also tried to authenticate the context through [Connection Service File](https://www.postgresql.org/docs/9.1/libpq-pgservice.html) and [Password File](https://www.postgresql.org/docs/9.1/libpq-pgpass.html) and mount them via `forward-credentials.yml` into the container.

mustafagok replied:
I have created connection in setUp, see: #8888

ashb commented on airflow/providers/amazon/aws/example_dags/example_s3_to_redshift.py:
```diff
@@ -0,0 +1,53 @@
+
+# [START howto_operator_s3_to_redshift_env_variables]
+S3_BUCKET = getenv("S3_BUCKET", "bucket")
+S3_KEY = getenv("S3_KEY", "key")
```
Why are we relying on env vars, rather Airflow Variables+templates? `{{ var.s3_key}}`

feluelle replied:
We have this in all of our system tests. So that we can easily pass them to the example dags. See https://github.com/apache/airflow/blob/master/TESTING.rst#environment-for-system-tests

feluelle replied:
I think it makes sense to switch to Airflow Variables+templates like Ash said and set them via env variables: https://airflow.apache.org/docs/stable/concepts.html?highlight=xcom#storing-variables-in-environment-variables. 

So we only need to prefix all env vars with `AIRFLOW_VAR`.
For example: set `AIRFLOW_VAR_S3_KEY=key` in `variables.env` and access them via `{{ var.s3_key}}` here. WDYT @potiuk @turbaszek ?

feluelle replied:
But how do we then define defaults in case these `var.`s are not set? Or should we make all vars required then?

potiuk replied:
I think this is fine and when we automate it, we should define files where we define ENV variables.

Using  Airflow Variables in example DAGs is a very bad idea. People use it as  - well - examples - and using airflow Variables in DAGs in this way is very bad from Database load point of view. Until we avoid parsing the DAGs by all entities frequently we should not use variables in this way (we even have best practice about it)  - basically every time the DAG is parsed, DB communication occurs and the extra connection is made - that is quite bad for scalability and not obvious at all, so if people take that literally as examples (which they did in the past) they will implement bad practices.

ashb commented on tests/test_utils/amazon_system_helpers.py:
```diff
@@ -82,6 +89,31 @@ def delete_aws_s3_bucket(cls, name: str) -> None:
+        :param key_name: name of the key to add
+        :param body: optional content of the key to add
+        """
+        cmd = ["aws", "s3api", "put-object", "--bucket", bucket_name, "--key", key_name]
```
Ummm. We have boto installed, let's use that instead of shelling out

feluelle replied:
True, but my thought is that we should not system-test our prod code with prod code? I mean theoretically if we change our code to use a different lib or so we also might need to change our test code.

feluelle replied:
Also some function we don't have in our hooks or operators. We then would need to mix hook, operator and boto3 calls. I think it would get confusing by having different abstraction layers.

feluelle replied:
@ashb Of course, terraform would be better to create resources but since we have `awscli` installed anyway, I think it is not bad to start like this and switch to terraform once we have it.

ashb commented on tests/test_utils/postgres_system_helpers.py:
```diff
@@ -0,0 +1,97 @@
+        from os import getenv
+        connection_string = getenv("AIRFLOW_CONN_POSTGRES_DEFAULT")
+        auth = ["psql", connection_string]
+        executor.execute_cmd(cmd=auth + ["-c", cmd])
```
Again, why are we shelling out rather than using DbAPiHook/psycopg2?

feluelle replied:
Same as above.

feluelle commented on tests/test_utils/postgres_system_helpers.py:
```diff
@@ -0,0 +1,97 @@
+
+
+@contextmanager
+def provide_postgres_table(schema_name, table_name, columns=None):
```
```suggestion
def provide_postgres_table(schema_name, table_name, columns):
```

xinbinhuang commented on tests/providers/amazon/aws/operators/test_s3_to_redshift_system.py:
```diff
@@ -0,0 +1,62 @@
+@pytest.fixture
+def provide_s3_bucket_with_key():
+    with provide_aws_s3_bucket(name=S3_BUCKET) as s3_bucket:
+        with NamedTemporaryFile() as file:
```
```suggestion
    with provide_aws_s3_bucket(name=S3_BUCKET) as s3_bucket, NamedTemporaryFile() as file:
```

feluelle commented on Dockerfile.ci:
```diff
@@ -308,26 +308,32 @@ RUN if [[ -n "${ADDITIONAL_PYTHON_DEPS}" ]]; then \
-    echo "docker run --rm -it -v \${HOST_HOME}/.aws:/root/.aws ${AWSCLI_IMAGE} \"\$@\"" \
-        > /usr/bin/aws && \
+    TERRAFORM_IMAGE="hashicorp/terraform:latest" && \
+    echo "#!/bin/bash" | tee -a /usr/bin/aws /usr/bin/az /usr/bin/bq /usr/bin/gcloud /usr/bin/gsutil /usr/bin/terraform && \
```
I added this line so python can run them properly - all bin's need to have a `#!/bin/bash` header.

feluelle commented on Dockerfile.ci:
```diff
@@ -308,26 +308,32 @@ RUN if [[ -n "${ADDITIONAL_PYTHON_DEPS}" ]]; then \
     echo "docker pull ${GCLOUD_IMAGE}" > /usr/bin/gsutil-update && \
-    chmod a+x /usr/bin/aws /usr/bin/az /usr/bin/bq /usr/bin/gcloud /usr/bin/gsutil
+    echo "docker run --rm -it -v \${HOST_HOME}/.aws:/root/.aws -v \${HOST_HOME}/.azure:/root/.azure -v \${HOST_HOME}/.config:/root/.config -v \${HOST_AIRFLOW_SOURCES}:/opt/airflow -w /opt/airflow --env-file <(env | grep TF) ${TERRAFORM_IMAGE} \"\$@\"" \
+        >> /usr/bin/terraform && \
```
Here I added all config dirs from aws, azure and google so terraform can use those to authenticate.

I also added `-v \${HOST_AIRFLOW_SOURCES}:/opt/airflow` to all of them so they can use airflow sources. For example terraform needs to access its `*.tf` and `*.tfvars` files.

`--env-file <(env | grep TF)` <- this is an interesting one. I did this to share all `TF*` env variables. So you can define those in `variables.env` from breeze and we do not need another one. Unfortunatly we can't just do `/files/airflow-breeze-config/variables.env` because this env file also has bash syntax in it - not only env var declarations.

feluelle commented on airflow/providers/amazon/aws/example_dags/example_s3_to_redshift.py:
```diff
@@ -41,13 +42,26 @@
+        sql=f'CREATE TABLE {REDSHIFT_TABLE}(Id int, Name varchar)',
+        postgres_conn_id='redshift_default',
+        task_id='preparation__create_table'
+    )
```
WDYT of having some `preparation` tasks? Not everything I can prepare with terraform.

turbaszek replied:
That's what we usually tried to do in GCP, make example that is self-sufficient so I'm ok with this approach. However, this requires few things:
- operator must be idempotent ("setup" steps in example must work always)
- consistency between what is created in example and what is deleted in test tear down

feluelle replied:
You are correct. But since I am `terraform destroy`ing the whole infra being used in `tearDown` I don't need to remove the s3 object and the redshift table, do I?

feluelle replied:
I do need to remove it, you are right. 👍 It should not depend on the `tearDown`.
And I will also make the setup steps idempotent. 👍

feluelle commented on scripts/ci/docker-compose/local-prod.yml:
```diff
@@ -40,4 +40,5 @@ services:
       - HOST_USER_ID
       - HOST_GROUP_ID
       - HOST_HOME=${HOME}
+      - HOST_AIRFLOW_SOURCES
```
@potiuk I added `HOST_AIRFLOW_SOURCES` here and set it in my `variables.env` file. WDYT?

feluelle commented on scripts/ci/docker-compose/local.yml:
```diff
@@ -62,6 +62,7 @@ services:
       - HOST_USER_ID
       - HOST_GROUP_ID
       - HOST_HOME=${HOME}
+      - HOST_AIRFLOW_SOURCES
```
and here.

feluelle commented on tests/test_utils/terraform.py:
```diff
@@ -0,0 +1,36 @@
+
+    def tearDown(self) -> None:
+        self.execute_cmd(["terraform", "plan", "-destroy", "-input=false", self.TERRAFORM_DIR])
+        self.execute_cmd(["terraform", "destroy", "-input=false", "-auto-approve", self.TERRAFORM_DIR])
```
I am not quite happy about it. This can probably be structured better.

turbaszek replied:
We can consider using/taking a look at https://github.com/beelit94/python-terraform

feluelle replied:
I checked this out, but didn't find that really useful to use an **extra package** only to wrap bash inside python. And it also [seems](https://github.com/beelit94/python-terraform/issues/85) to not be compatible to the latest terraform 0.12.x releases.

turbaszek replied:
Then I would suggest to add `init`, `plan`, `apply` as methods/functions to limit the number of replicated code using `self.execute_cmd`. WDYT?

feluelle replied:
I could do that.

The real issue I am having with this, is also that `super().setUp()` and `super().tearDown()` are not explicit enough. Terraform is very much hidden behind them. You can easily forget to call those when you overwrite those which for `super().tearDown()` could be dramatically. But maybe we could add a static check that checks that?!

feluelle reviewed (commented):
First try (this works 🎉 ) - I switched to terraform for providing infra. WDYT? @potiuk @ashb @mik-laj @turbaszek @kaxil

turbaszek commented on tests/providers/amazon/aws/operators/test_s3_to_redshift_system.py:
```diff
@@ -0,0 +1,44 @@
+        schema = self.get_tf_output("redshift_database_name"),
+        login = self.get_tf_output("redshift_master_username"),
+        password = self.get_tf_output("redshift_master_password")
+        db.merge_conn(Connection("redshift_default", "postgres", host, login, password, schema, port))
```
Is this connection deleted after test?

feluelle replied:
Nope. That's a TODO. Thanks 👍

feluelle committed:
Add S3ToRedshift example dag and system test

- add howto docs for S3ToRedshift example dag
- add terraform which runs terraform CLI commands in an isolated docker container

NOTE: This system test uses terraform to provide the infrastructure needed to run this example dag.
